Firestore/core/src/firebase/firestore/remote/datastore.mm

```diff
@@ -292,6 +308,44 @@ void LogGrpcCallFinished(absl::string_view rpc_name,
   active_calls_.erase(found);
 }
 
+bool Datastore::IsAbortedError(const Status& error) {
+  return error.code() == FirestoreErrorCode::Aborted;
+}
+
+bool Datastore::IsPermanentError(const Status& error) {
+  switch (error.code()) {
+    case FirestoreErrorCode::Cancelled:
+    case FirestoreErrorCode::Unknown:
+    case FirestoreErrorCode::DeadlineExceeded:
+    case FirestoreErrorCode::ResourceExhausted:
+    case FirestoreErrorCode::Internal:
+    case FirestoreErrorCode::Unavailable:
+      // Unauthenticated means something went wrong with our token and we need
+      // to retry with new credentials which will happen automatically.
+    case FirestoreErrorCode::Unauthenticated:
+      return false;
+    case FirestoreErrorCode::InvalidArgument:
+    case FirestoreErrorCode::NotFound:
+    case FirestoreErrorCode::AlreadyExists:
+    case FirestoreErrorCode::PermissionDenied:
+    case FirestoreErrorCode::FailedPrecondition:
+    case FirestoreErrorCode::Aborted:
+      // Aborted might be retried in some scenarios, but that is dependant on
+      // the context and should handled individually by the calling code.
+      // See https://cloud.google.com/apis/design/errors
+    case FirestoreErrorCode::OutOfRange:
+    case FirestoreErrorCode::Unimplemented:
+    case FirestoreErrorCode::DataLoss:
+      return true;
+    default:
```

Comment:
Our compile flags will ensure that all cases are handled, so you can drop the `default`. However, you still need to worry about the error code having values that aren't valid enum values.

Proposal:
1) drop the default.
2) Add the following, outside the switch statement:
```
log("unexpected status error code: %i", error.code());
return true;
```

This will have the same effect as your code (modulo the log) and has the added "benefit" (or maybe drawback, depending on your point of view) of failing to compile if a new enum value is added to the FirestoreErrorCode enum. (Which is exceptionally unlikely to occur.)

Optional.